templates/control/edit_apikeys.html

```diff
@@ -78,16 +88,65 @@
         <button class="button negative" style="float: right;">Revoke Selected Consumers</button>
       $else:
         <p style="padding-top: 1em;">You have no OAuth2 consumers!</p>
+    </form>
 
+    <h3>Register OAuth2 Application</h3>
+    <form action="" method="POST" class="clear">
+      $:{CSRF()}
+      <p style="padding-top: 1em;">
+        If you are a developer, you may register an application as an OAuth2 consumer.
+        You can then use your client id and client secret to request authorization tokens from
+        Weasyl. Your client secret MUST be stored securely and should not be accessible
+        by your application's users under any circumstance. For more information see
+        <a href="https://projects.weasyl.com/weasylapi/#oauth2-endpoints">
+          https://projects.weasyl.com/weasylapi/#oauth2-endpoints</a>
+      </p>
+      <label for="client-name">Application Name</label>
+      <input id="client-name" type="text" name="client-name" class="input" required />
 
+      <label for="client-homepage">Application Homepage</label>
+      <input id="client-homepage" type="url" name="client-homepage" class="input" placeholder="optional" />
+
+      <label for="redirect-uris">Redirect URIs</label>
+      <input id="redirect-uris" type="url" name="redirect-uris" class="input" required
+          placeholder="separate multiple URIs with a space"/>
+
+      <label>Scopes</label>
+      $for scope in oauth_scopes:
+        <label class="input-checkbox">
+          <input type="checkbox" name="client-scopes" value="${scope['name']}">
+          <strong>${scope['name']}</strong> - ${scope['description']}
+        </label>
+
+      <button name="create-oauth2-client" value="y" style="float:right;margin-top:1em"
+              class="button positive">Register Application</button>
     </form>
-  </div>
 
-  <h3>How to use API keys</h3>
-  <p>
-    To connect to Weasyl using an API key, set the value of
-    the <tt>X-Weasyl-API-Key</tt> HTTP header to the API key. This will
-    override any cookie-based authentication.
-  </p>
+    <h3>Manage OAuth2 Applications</h3>
+      $if my_applications:
+        <form action="" method="POST" class="clear">
+        $:{CSRF()}
+        $for app in my_applications:
+          <label class="input-checkbox">
+            <input type="checkbox" name="clientid" value="${app.clientid}" />
+            $if app.homepage:
+              <a href="${app.homepage}"><strong>${app.description}</strong></a>
```

Comment:
Why? `target="_blank"` is considered irritating to many people and we’d probably want to guard against `window.opener` as a matter of good practice (even though it’d effectively be people attacking themselves here, it’s not too far-fetched to think that we might display integrated apps’ home page somewhere else).